docs/contributor/adr/002-constructor-injection-and-composition-functions.md

```diff
@@ -0,0 +1,96 @@
+# ADR 002 - Constructor Injection and Composition Functions
+
+## Status
+
+Accepted
+
+## Context
+
+It is agreed that [Dependency Inversion](https://medium.com/@inzuael/solid-dependency-inversion-principle-part-5-f5bec43ab22e) is a key principle for solid architecture.
+Once a consumer abstracted their dependencies properly, the key question is how to fill the abstraction with a concrete implementation.
+
+It has to be decided how and where to resolve dependencies.
+
+## Decision
+
+It is decided that a consumer must have its dependencies injected via constructor.
+Constructors should follow the naming pattern `New<struct name>`.
+Stable dependencies of a consumer are resolved in a *composition function*.
+Composition functions live in a separate pacakge at `cmd/composition`.
+Composition functions should follow the naming pattern `Compose<struct name>`.
+Composition functions may build dependencies of the unit to compose themselves or have them injected from outside (depends on the kind of dependency).
+Composition functions do NOT return errors but instead log them and exit the startup process.
+From `main`, the top-layer composition functions are called only exposing the top-layer dependencies.
+Within the composition functions, other composition functions from the downstream layers may be called.
+
+## Consequence
+
+Consumers do not build or know the concrete implementations of their abstracted dependencies.
+The `main` function is kept free from resolving the detailed dependency tree.
+The detailed dependency tree is resolved via the compositions functions under `cmd/composition`.
+
+As the above decisions are applied, they may be continuously refined and extended.
+
+### Example
+
+```go
+// /cmd/composition/service/webhook/skr_webhook_manager.go
+package webhook
+
+import (
+  "github.com/kyma-project/lifecycle-manager/internal/watcher"
+)
+
+func ComposeSkrWebhookManager(logger logr.Logger, flagVar *flags.FlagVar) *watcher.SkrWebhookManager {
+  certManager := composeCertManager(logger, flagVar)
```

Comment:
I prefer not to use deeply nested Composition functions. For components like `certManager` as stable dependencies, it's more straightforward to construct them directly rather than introducing an extra layer of composition. Adding such levels only complicates usage unnecessarily. All Composition functions should be public and only invoked in `main`, as they are intended for components with stable dependencies. Not all consumers will need or use these composition functions.